tower-http/src/services/fs/parse_range.rs

```diff
@@ -0,0 +1,309 @@
+                        continue;
+                    }
+                    return ParsedRangeHeader::Malformed(format!(
+                        "Range: {} is not acceptable, end of range not parseable.",
+                        range
+                    ));
+                }
+                if let Ok(start) = start.parse::<u64>() {
+                    if end == "" {
+                        ranges.push(RangeInclusive::new(start, file_size_bytes - 1));
+                        continue;
+                    }
+                    if let Ok(end) = end.parse::<u64>() {
+                        if end >= file_size_bytes {
+                            return ParsedRangeHeader::Unsatisfiable(format!(
+                                "Range: {} is not satisfiable, end of range exceeds file boundary.",
+                                range
+                            ));
+                        }
+                        ranges.push(RangeInclusive::new(start, end));
+                        continue;
+                    }
+                    return ParsedRangeHeader::Malformed(format!(
+                        "Range: {} is not acceptable, end of range not parseable.",
+                        range
+                    ));
+                }
+                return ParsedRangeHeader::Malformed(format!(
+                    "Range: {} is not acceptable, start of range not parseable.",
+                    range
+                ));
+            }
+            return ParsedRangeHeader::Malformed(format!(
+                "Range: {} is not acceptable, range does not contain any dashes.",
+                range
+            ));
+        }
+    } else {
+        return ParsedRangeHeader::Malformed(format!(
+            "Range: {} is not acceptable, range does not start with 'bytes='",
+            range
+        ));
+    }
+    if ranges.is_empty() {
+        panic!("Programming error parsing range {}", range)
+    } else {
+        if ranges.len() == 1 {
+            ParsedRangeHeader::Range(ranges)
+        } else if !overlaps(&ranges) {
+            ParsedRangeHeader::Range(ranges)
+        } else {
+            return ParsedRangeHeader::Unsatisfiable(format!(
+                "Range header: {} is not satisfiable, ranges overlap",
+                range
+            ));
+        }
+    }
+}
+
+fn overlaps(ranges: &Vec<RangeInclusive<u64>>) -> bool {
+    let mut bounds = Vec::new();
+    for range in ranges {
+        bounds.push((range.start(), range.end()));
+    }
+    for i in 0..bounds.len() {
+        for j in i + 1..bounds.len() {
+            if bounds[i].0 <= bounds[j].1 && bounds[j].0 <= bounds[i].1 {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+fn split_once<'a>(s: &'a str, pat: &'a str) -> Option<(&'a str, &'a str)> {
+    let mut iter = s.split(pat);
+    let left = iter.next();
+    left.and_then(|l| iter.next().map(|r| (l, r)))
+}
```

Comment:
`?` works on `Option` so this should work:

```suggestion
fn split_once<'a>(s: &'a str, pat: &'a str) -> Option<(&'a str, &'a str)> {
    let mut iter = s.split(pat);
    let left = iter.next()?;
    let right = iter.next()?;
    Some((left, right))
}
```